### DISCUSSION TITLE
为什么用户控件中的CheckBox的IsChecked需要绑定一个类的布尔才能通知更新

### DISCUSSION THREAD

**Discussion by ZhuYajunFly** [BYJRK/DotNet-Discussions#341] [MVVM]
为什么用户控件中的CheckBox的IsChecked需要绑定一个类的布尔才能通知更新，直接绑定一个布尔更新不了，从前台通知到代码。都有[ObservableProperty]特性。我看了类型是Bool? 但是我声明一个Bool? 变量也不能前台通知到代码。

**绑定类的布尔可以更新**
![image](https://github.com/user-attachments/assets/961ead62-8166-4ac9-9be4-0b6add0d9328)
![image](https://github.com/user-attachments/assets/b079db32-4403-4dff-b612-da8f99bd0a90)

**直接绑定布尔不更新**
![image](https://github.com/user-attachments/assets/fe95fab2-6c38-44a9-b423-b1dd0436eac1)
![image](https://github.com/user-attachments/assets/785d55be-1b5e-4a65-9204-21fd24d212a5)

**Comment by arcanix601** [comment]
[WpfApp1.zip](https://github.com/user-attachments/files/16871616/WpfApp1.zip)
我这个都可以通知到，你参考一下吧

**Reply by ZhuYajunFly** [reply]
> [WpfApp1.zip](https://github.com/user-attachments/files/16871616/WpfApp1.zip) 我这个都可以通知到，你参考一下吧

把CheckBox放到UserControl用户控件里面，通过绑定呢，我是这样用的

**Reply by ZhuYajunFly** [reply]
> [WpfApp1.zip](https://github.com/user-attachments/files/16871616/WpfApp1.zip) 我这个都可以通知到，你参考一下吧

怪哉  我刚刚写了一个简单的demo 然后UserControl用户控件指定双向绑定后，又可以了。但是我今天下午在另一个项目，测试了半个小时，都不成功，换成类包装就行。怪了

**Reply by ZhuYajunFly** [reply]
> [WpfApp1.zip](https://github.com/user-attachments/files/16871616/WpfApp1.zip) 我这个都可以通知到，你参考一下吧

可能是单布尔我没有加 IsChecked="{Binding IsAutoSave, Mode=TwoWay}"中的Mode=TwoWay，所以不能通知到代码。
可能是类中的布尔不加Mode=TwoWay，可以界面通知到代码。

**Reply by ZhuYajunFly** [reply]
> [WpfApp1.zip](https://github.com/user-attachments/files/16871616/WpfApp1.zip) 我这个都可以通知到，你参考一下吧

感谢您哈

**Reply by ZhuYajunFly** [reply]
> [WpfApp1.zip](https://github.com/user-attachments/files/16871616/WpfApp1.zip) 我这个都可以通知到，你参考一下吧

单布尔用户控件不双向绑定确实有点问题，我上传一个demo。
**来回点击两圈就能发现问题**
[WpfApp1.zip](https://github.com/user-attachments/files/16873113/WpfApp1.zip)

**Reply by arcanix601** [reply]
> > [WpfApp1.zip](https://github.com/user-attachments/files/16871616/WpfApp1.zip) 我这个都可以通知到，你参考一下吧
> 
> 单布尔用户控件不双向绑定确实有点问题，我上传一个demo。 **来回点击两圈就能发现问题** [WpfApp1.zip](https://github.com/user-attachments/files/16873113/WpfApp1.zip)

确实是TwoWay的问题，可以这样写依赖属性，绑定时默认TwoWay
```csharp
        public static readonly DependencyProperty IsAutoSaveNoProperty =
            DependencyProperty.Register("IsAutoSaveNo",    
                typeof(bool), 
                typeof(UC01), 
                new FrameworkPropertyMetadata(true, FrameworkPropertyMetadataOptions.BindsTwoWayByDefault));
```

**Reply by ZhuYajunFly** [reply]
> > > [WpfApp1.zip](https://github.com/user-attachments/files/16871616/WpfApp1.zip) 我这个都可以通知到，你参考一下吧
> > 
> > 
> > 单布尔用户控件不双向绑定确实有点问题，我上传一个demo。 **来回点击两圈就能发现问题** [WpfApp1.zip](https://github.com/user-attachments/files/16873113/WpfApp1.zip)
> 
> 确实是TwoWay的问题，可以这样写依赖属性，绑定时默认TwoWay
> 
> ```cs
>         public static readonly DependencyProperty IsAutoSaveNoProperty =
>             DependencyProperty.Register("IsAutoSaveNo",    
>                 typeof(bool), 
>                 typeof(UC01), 
>                 new FrameworkPropertyMetadata(true, FrameworkPropertyMetadataOptions.BindsTwoWayByDefault));
> ```

是的，大佬用另一个重载方法，确实在XAML两处都不用去写Mode=TwoWay了。昨天真的是纳了闷了，然后又到了下班点，就没有去测试Mode=TwoWay，我当时的想法是类中的布尔能通知，那我最直接的布尔应该也能呀。
